src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java

```diff
@@ -0,0 +1,169 @@
+import org.sagebionetworks.web.client.security.AuthenticationController;
+import org.sagebionetworks.web.client.view.EmailInvitationView;
+import org.sagebionetworks.web.client.view.users.RegisterWidget;
+import org.sagebionetworks.web.client.widget.entity.controller.SynapseAlert;
+import org.sagebionetworks.web.shared.NotificationTokenType;
+
+import com.google.common.util.concurrent.FluentFuture;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.gwt.activity.shared.AbstractActivity;
+import com.google.gwt.event.shared.EventBus;
+import com.google.gwt.user.client.ui.AcceptsOneWidget;
+
+public class EmailInvitationPresenter extends AbstractActivity implements EmailInvitationView.Presenter, Presenter<EmailInvitation> {
+	private String encodedMISignedToken;
+	private EmailInvitationView view;
+	private RegisterWidget registerWidget;
+	private SynapseJavascriptClient jsClient;
+	private SynapseFutureClient futureClient;
+	private SynapseAlert synapseAlert;
+	private AuthenticationController authController;
+	private PlaceChanger placeChanger;
+
+	@Inject
+	public EmailInvitationPresenter(EmailInvitationView view,
+									RegisterWidget registerWidget,
+									SynapseJavascriptClient jsClient,
+									SynapseFutureClient futureClient,
+									SynapseAlert synapseAlert,
+									AuthenticationController authController,
+									GlobalApplicationState globalApplicationState) {
+		this.view = view;
+		this.registerWidget = registerWidget;
+		this.jsClient = jsClient;
+		this.futureClient = futureClient;
+		this.synapseAlert = synapseAlert;
+		this.authController = authController;
+		this.placeChanger = globalApplicationState.getPlaceChanger();
+	}
+
+	@Override
+	public void start(AcceptsOneWidget acceptsOneWidget, EventBus eventBus) {
+		acceptsOneWidget.setWidget(view);
+	}
+
+	@Override
+	public void setPlace(EmailInvitation place) {
+		view.setPresenter(this);
+		view.clear();
+		encodedMISignedToken = place.toToken();
+
+		futureClient.hexDecodeAndDeserialize(NotificationTokenType.EmailInvitation.name(), encodedMISignedToken)
+			.transformAsync(
+					token -> jsClient.getMembershipInvitation((MembershipInvtnSignedToken) token),
+					directExecutor()
+			).addCallback(
+					new FutureCallback<MembershipInvtnSubmission>() {
+						@Override
+						public void onSuccess(MembershipInvtnSubmission mis) {
+							if (!authController.isLoggedIn()) {
+								initializeView(mis);
+							} else {
+								bindInvitationToAuthenticatedUser(mis.getId());
+							}
+						}
+
+						@Override
+						public void onFailure(Throwable throwable) {
+							synapseAlert.handleException(throwable);
+						}
+					},
+					directExecutor()
+			);
+	}
+
+	private void bindInvitationToAuthenticatedUser(final String misId) {
+		jsClient.getInviteeVerificationSignedToken(misId)
+			.transformAsync(
+					jsClient::updateInviteeId,
```

Comment:
I need your help understanding this.
"jsClient::updateInviteeId"
Could you explain this syntax?
Calling getInviteeVerificationSignedToken with the MembershipInvtnSubmission ID binds it to the currently authenticated user?

Reply:
This is passing updateInviteeId as a method reference to transformAsync. It's equivalent to "token -> jsClient.updateInviteeId(token)". Since the first parameter of transformAsync (AsyncFunction) is a functional interface, it can take in a method reference with one parameter (like above), a closure, or an instance of an AsyncFunction class.

Reply:
Please change to lambda